### PR TITLE
Add LightningTransformer demo for PyTorch Lightning in `lightning.pytorch.demos.transformer`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

- adds LightningTransformer demo for PyTorch Lightning

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: docs/source-pytorch/common/trainer.rst [hunk 1]
```diff
@@ -158,7 +169,7 @@ Example::
 
     seed_everything(42, workers=True)
     # sets seeds for numpy, torch and python.random.
-    model = Model()
+    model = LightningTransformer(vocab_size=dataset.vocab_size)
```
Comment (awaelchli) [inline docs/source-pytorch/common/trainer.rst:172]: The main message here is reproducibility, but the details of the model here distract. The previous line was clearer.

File: docs/source-pytorch/common/trainer.rst [hunk 2]
```diff
@@ -1064,15 +1075,15 @@ With :func:`torch.inference_mode` disabled, you can enable the grad of your mode
 
 .. code-block:: python
 
-    class LitModel(LightningModule):
+    class LightningTransformer(pl.LightningModule):
         def validation_step(self, batch, batch_idx):
-            preds = self.layer1(batch)
+            preds = self.model(input.view(1, -1), target.view(1, -1))
```
Comment (awaelchli) [inline docs/source-pytorch/common/trainer.rst:1080]: It is unclear to me why we reshape the input like this.

File: src/lightning/pytorch/demos/lightning_transformer.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        return self.model(input.view(1, -1), target.view(1, -1))
+
+    def training_step(self, batch: Tuple[Tensor, Tensor], batch_idx: int) -> Tensor:
+        input, target = batch
+        output = self.model(input.view(1, -1), target.view(1, -1))
+        loss = torch.nn.functional.nll_loss(output, target.view(-1))
+        return loss
+
+    def predict_step(self, batch: Tuple[Tensor, Tensor]) -> Tensor:
+        return self(batch)
```
Comment (awaelchli) [inline src/lightning/pytorch/demos/lightning_transformer.py:26]: This is already the default. Is the override intentional?

File: src/lightning/pytorch/demos/lightning_transformer.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+from typing import Tuple
+
+import torch
+from torch import Tensor
+
+import lightning.pytorch as pl
+from lightning.pytorch.demos.transformer import Transformer
```
Comment (carmocca) [inline src/lightning/pytorch/demos/lightning_transformer.py:7]: I would define this LightningModule together with the nn.Module Transformer

Comment (JustinGoheen) [reply src/lightning/pytorch/demos/lightning_transformer.py]: moved in [05459ca](https://github.com/Lightning-AI/lightning/pull/18452/commits/05459caf327ed834c88f61e1dbefcc4a0a4b240f)

File: src/lightning/pytorch/demos/lightning_transformer.py (hunk 3 above)
Comment (JustinGoheen) [reply src/lightning/pytorch/demos/lightning_transformer.py]: removed in [05459ca](https://github.com/Lightning-AI/lightning/pull/18452/commits/05459caf327ed834c88f61e1dbefcc4a0a4b240f)

File: docs/source-pytorch/common/trainer.rst (hunk 2 above)
Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst]: updated in [05459ca](https://github.com/Lightning-AI/lightning/pull/18452/commits/05459caf327ed834c88f61e1dbefcc4a0a4b240f)

File: docs/source-pytorch/common/trainer.rst (hunk 1 above)
Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst]: changed in [05459ca](https://github.com/Lightning-AI/lightning/pull/18452/commits/05459caf327ed834c88f61e1dbefcc4a0a4b240f)

File: docs/source-pytorch/common/trainer.rst [hunk 5]
```diff
@@ -33,10 +33,22 @@ This is the basic use of the trainer:
 
 .. code-block:: python
 
-    model = MyLightningModule()
+    import lightning.pytorch as pl
+    from lightning.pytorch import Trainer
```
Comment (awaelchli) [inline docs/source-pytorch/common/trainer.rst:37]: I suggest to reduce this import
```suggestion
```
and below use

pl.Trainer()

File: docs/source-pytorch/common/trainer.rst [hunk 6]
```diff
@@ -33,10 +33,22 @@ This is the basic use of the trainer:
...
+
+    from lightning.pytorch.demos import WikiText2
+    from lightning.pytorch.demos import LightningTransformer as MyLightningModule
```
Comment (awaelchli) [inline docs/source-pytorch/common/trainer.rst:40]: I suggest to combine
```suggestion
    from lightning.pytorch.demos import LightningTransformer as MyLightningModule, WikiText2
```
(unless it makes the line too long for the code block)

**Review by awaelchli — APPROVED** [review]

File: docs/source-pytorch/common/trainer.rst (hunk 5 above)
Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst:37]: I had done this previously and reverted because there are some many instances of `trainer = Trainer()` on the page that it makes sense to leave this as is until I can update the docstrings that also have `trainer = Trainer()` .

Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst]: implemented in [7cf7aff](https://github.com/Lightning-AI/lightning/pull/18452/commits/7cf7aff92dddb21c33efac1902289d313658a7f4)

File: docs/source-pytorch/common/trainer.rst [hunk 7]
```diff
@@ -33,10 +33,20 @@ This is the basic use of the trainer:
 
 .. code-block:: python
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:34]: IMO this loses the initial purpose of the snippet which was to "show the basic use of the Trainer". The user now has to understand what is wikitext, random_split, a transformer etc

File: docs/source-pytorch/common/trainer.rst [hunk 8]
```diff
@@ -1064,15 +1074,15 @@ With :func:`torch.inference_mode` disabled, you can enable the grad of your mode
...
 
-    class LitModel(LightningModule):
+    class LightningTransformer(pl.LightningModule):
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:1077]: Since you already import a class with this name in the snippet above, why not keep this other class with a different name?

**Review by carmocca — APPROVED** [review]
I don't fully agree with the design direction but the changes look technically correct to me. Approving since it's not for me to decide the docs direction :)

File: docs/source-pytorch/common/trainer.rst (hunk 7 above)
Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst:34]: My thought process here is that I can use Adrian's transformer and dataset as a consistent example throughout docs in order to replace the encoder-decoder as the base example.

File: docs/source-pytorch/common/trainer.rst (hunk 8 above)
Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst:1077]: I just wanted to be consistent in the model names used throughout.

File: docs/source-pytorch/common/trainer.rst (hunk 7 above)
Comment (carmocca) [reply docs/source-pytorch/common/trainer.rst:34]: Replacing the encoder-decoder example is a good idea. But the previous code wasn't using that but skeletons that don't implement anything

Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst:34]: reverted in [426e8d5](https://github.com/Lightning-AI/lightning/pull/18452/commits/426e8d5c02e99798fbe473527d3e64023b245451)

Comment (carmocca) [reply docs/source-pytorch/common/trainer.rst:34]: Note that none of the examples in this file use the encoder-decoder example. Will you revert them too?

File: src/lightning/pytorch/__init__.py [hunk 9]
```diff
@@ -23,6 +23,7 @@
     _logger.propagate = False
 
 from lightning.fabric.utilities.seed import seed_everything  # noqa: E402
+from lightning.pytorch import demos  # noqa: E402
```
Comment (awaelchli) [inline src/lightning/pytorch/__init__.py:26]: See https://github.com/Lightning-AI/lightning/pull/18451#discussion_r1319560055

Either we do this or change the import in the example. cc @Borda

Comment (carmocca) [reply src/lightning/pytorch/__init__.py:26]: I strongly suggest that this isn't imported by default. It's an unnecessary overhead for 99.99% of users.

File: docs/source-pytorch/common/trainer.rst (hunk 7 above)
Comment (awaelchli) [reply docs/source-pytorch/common/trainer.rst:34]: Yeah we need to put the import at the top here otherwise to the reader `LightningTransformer` is just a name without definition anywhere.

Comment (JustinGoheen) [reply docs/source-pytorch/common/trainer.rst]: @awaelchli there shouldn't be anymore uses of `LightningTransformer` as of [6671aa6](https://github.com/Lightning-AI/lightning/pull/18452/commits/6671aa6215041d1f8598854e39fd7a5c5330cb69)

**Comment by JustinGoheen** [comment]
Rather than updating common/trainer, this PR now only adds `LightningTransformer` to `lightning.pytorch.demos.transformer`.
